### ISSUE TITLE
Execute in Jetson Nano 4GB

### ISSUE THREAD

**Issue by irvinmarceloc** [lojzezust/WaSR-T#1]
Hi guys, 
I had problem when execute the command 
pip install -r requirements.txt
With the following output

_**(env) jetbot@nano-4gb-jp45:~ /WaSR-T$ pip install -r requirements.txt 
Collecting torch==1.8.1 (from -r requirements.txt (line 1))
  Could not find a version that satisfies the requirement torch==1.8.1 (from -r requirements.txt (line 1)) (from versions: )
No matching distribution found for torch==1.8.1 (from -r requirements.txt (line 1))
(env) jetbot@nano-4gb-jp45:~ /WaSR-T$ pip install torch
Collecting torch
  Could not find a version that satisfies the requirement torch (from versions: )
No matching distribution found for torch**_

I have a virtual enviroment with python 3.8 install on ubuntu 18.04 for arm arch.

**Comment by lojzezust** [comment]
Hi! That seems to be a problem with PyTorch installation through pip. Try following the [official PyTorch installation instructions](https://pytorch.org/get-started/locally/) for your system. If that doesn't work you could also try to install using conda.

**Comment by irvinmarceloc** [comment]
Hi thanks. I fixed it by updating pip.
I'm trying to run this project on jetson nano, however as much as I've managed it is quite slow, taking up to 40 seconds per image. Is that normal?

**Comment by lojzezust** [comment]
Great to hear!

Jetson Nano is not a very powerful device, but 40s seems extremely slow. Are you sure you are running the inference on the GPU and are using the *sequential mode* of WaSR-T? 

The inference speed of WaSR-T (in sequential mode) is about 9-10 FPS on RTX 2080 Ti and takes about 2.5 GB of memory. Considering Jetson Nano is capable of around 500 GFLOPS, the inference speed should be around 0.5 FPS, ignoring other scaling factors such as CPU and time spent loading images into memory. I am not too familiar with embedded devices and the bottlenecks that might occur when running inference there.

I would suggest using TensorRT to optimize the model for inference. However, to run WaSR-T on this device fully in real-time, you would probably also need some modifications to the architecture, such as using a faster backbone (e.g. MobileNet or smaller ResNet).

**Comment by irvinmarceloc** [comment]
The truth is that I'm not sure if it's with the GPU.
Should it be enough to define the environment variable? For example
`export CUDA_VISIBLE_DEVICES=128`
Or should I do something else?
Maybe you are right and I am running it on CPU, while running the inference I am seeing excessive CPU usage.

What should I take into account to opt for MobileNet or ResNet18?
Thank you very much in advance for your help.

**Comment by lojzezust** [comment]
First check if CUDA is set up properly and detects your GPU. If you run `nvidia-smi` in the terminal in your environement you should see the list of GPUs.

I expect a single GPU is installed on the machine. If that's the case you don't need to set `CUDA_VISIBLE_DEVICES` or you can set `export CUDA_VISIBLE_DEVICES=0` to only show the first device.

Then test in python if torch can see your GPU by using `torch.cuda.is_available()`. The prediction code uses this check to determine whether to run on cuda or on cpu.

---

See [`wasr_t.wasr_t.wasr_temporal_resnet101()`](https://github.com/lojzezust/WaSR-T/blob/master/wasr_t/wasr_t.py#L17) to see how the backbone is built. WaSR-T requires outputs at 4 different stages of the backbone. You can build a model with a different backbone in a similar way. Note that `WaSRTDecoder` expects the number of output channels of ResNet-101. If you use a smaller backbone, the output channel dimensions will be smaller and you will need to update the layer parameters in the decoder accordingly.

You will also need to re-train the new network. Follow the instructions [here](https://github.com/lojzezust/WaSR-T#model-training).

**Comment by irvinmarceloc** [comment]
Hello, sorry, it took me a while to answer. I didn't have gpu enabled, I had to download a whl installer from the nvidia page for this to work.
I will try to document well how to implement this project in jetson nano to share it with the community

**Comment by lojzezust** [comment]
Great, I'm glad to hear the good news! What kind of performance do you get on Jetson Nano?

If you decide to document this process, I would be more than happy to include it in the project. I believe it would be of great interest and value to the community. Let me know if you need any help.

**Comment by irvinmarceloc** [comment]
Hi, thanks for the help. Performance has improved.
It went from 40 seconds to 9 seconds per image on average.
I will try changing the model and using the same dataset. Or something you recommend me to do.
I would like to collaborate in the project including the documentation when I finish the implementation that I am doing.

**Comment by irvinmarceloc** [comment]
It currently looks like this
![It currently looks like this](https://user-images.githubusercontent.com/64157096/197391216-9fdd611c-a720-4d7b-83a5-ac6d5f72ca47.png)
Previously when running the inference the CPU looked like this
![Previously when running the inference the CPU looked like this](https://user-images.githubusercontent.com/64157096/197391311-c1fae3fb-cd84-4e98-bf44-43c343da1b78.png)

**Comment by irvinmarceloc** [comment]
Hello me again, I want to reduce the time of 0.125 FPS and for that I check if all the libraries support GPU and I found that they do not, I installed cv2 with GPU and now I have this error

jettson@jettson-desktop:~/WaSR-T/path/to/model$ python3.6 /home/jettson/WaSR-T/predict_sequential.py --sequence-dir /home/jettson/WaSR-T/examples/sequence --weights /home/jettson/WaSR-T/path/to/model/wasrt_mastr1478.pth --output-dir /home/jettson/WaSR-T/output/predictions
Namespace(fp16=False, gpus=-1, hist_len=5, output_dir='/home/jettson/WaSR-T/output/predictions', sequence_dir='/home/jettson/WaSR-T/examples/sequence', weights ='/home/jettson/WaSR-T/path/to/model/wasrt_mastr1478.pth')
Processing frames: 0%| | 0/85 [00:00<?, ?it/s]/home/jettson/.local/lib/python3.6/site-packages/torchvision-0.9.0-py3.6-linux-aarch64.egg/torchvision /transforms/functional.py:365: UserWarning: Argument interpolation should be of type InterpolationMode instead of int. Please, use InterpolationMode enum.
  "Argument interpolation should be of type InterpolationMode instead of int. "
Processing frames: 1%|█ | 1/85 [01:19<1:51:52, 79.92s/it]
Traceback (most recent call last):
  File "/home/jettson/WaSR-T/predict_sequential.py", line 93, in <module>
    main()
  File "/home/jettson/WaSR-T/predict_sequential.py", line 89, in main
    run_inference(args)
  File "/home/jettson/WaSR-T/predict_sequential.py", line 83, in run_inference
    predict_sequence(predictor, args.sequence_dir, output_dir)
  File "/home/jettson/WaSR-T/predict_sequential.py", line 70, in predict_sequence
    probs = predictor.predict_batch(features)
  File "/home/jettson/WaSR-T/wasr_t/inference.py", line 38, in predict_batch
    out = res['out'].cpu().detach()
RuntimeError: CUDA error: the launch timed out and was terminated

**Comment by lojzezust** [comment]
Hey! I'm not sure about this error. However, I believe cv2 is not necessary for running this network. It was mainly used for data augmentation during training. The data loading should and probably already does work without it.

As for recommendations for improving the runtime. There are several simple things you can try first:
- Using a smaller backbone (e.g. ResNet-101 should be relatively simple to replace with ResNet-50)
- Using half precision (FP16). Check if Jetson Nano supports this. The code already supports training in fp16 (`--precision 16`). Similar with inference.
- Using TensorRT to optimize the network for your device.

All these should improve the performance. Note that the first two will degrade the accuracy somewhat.

If this is still not fast enough, I recommend using a lightweight segmentation network (e.g. BiSeNet or TopFormer) and adding the Temporal Context Module of WaSR-T inside. This may require a bit of tinkering. 

If you decide to go this route, I recommend that you first test the network directly without changes. It may work well enough for your use case without requiring to spend a lot of time to implement TCM.

**Comment by irvinmarceloc** [comment]
I think I'm going to have to change the model, to be able to execute the algorithm as it is, use these commands that disable the watchdog that kills the processes.
```
sudo -s 
$ echo N > /sys/kernel/debug/gpu.0/timeouts_enabled 
$ sudo nvpmodel -m 0
$ sudo jetson_clocks
```
What characteristics did the computer have that they used to obtain the 10FPS in image processing? I read that information in the wasr-t paper.
Thanks for everything.

**Comment by irvinmarceloc** [comment]
Each image took approximately 7.7 seconds on average to process.

**Comment by lojzezust** [comment]
The system used is a compute server with an Intel Xeon Silver 4214 CPU. We used a single NVIDIA RTX 2080 Ti GPU for processing.

We also tested it on a local system with an NVIDIA GTX 1660 Ti, Intel Core i7-6700K CPU and 16GB of RAM, where it runs at around 3-4 FPS.

**Comment by playertr** [comment]
Heads up, @lojzezust : I'm also porting the model to a Jetson 4GB, and I was attempting to run inference on a video that had too high of a resolution. In the code right now, I don't think there is any automatic resizing. If you want to do real-time inference, you might need to downsample your input images, by, for example, applying a 384x512 `torchvision.transforms.Resize` block [where you ingest the data](https://github.com/lojzezust/WaSR-T/blob/efa4699128ff88b7332df9876523944fdbb5f80a/wasr_t/data/transforms.py#L68).

**Comment by irvinmarceloc** [comment]
Oh thanks for continuing.

How do you transfer the model? . Do you have a step by step for that?. I'm new to all this and some concepts still confuse me.
I tried to use nvidia DIGITS but I didn't have any compatible computer.

Last year I was very interested in being able to make this model work in jetson nano and although I got it I couldn't make it work in real time.

Regards, Irvin.

**Comment by playertr** [comment]
@irvinmarceloc Pardon the late response.

I encountered trouble shifting my [fork](https://github.com/lojzezust/WaSR-T/compare/master...playertr:WaSR-T:master) from my desktop computer with an RTX 2060 card to the Jetson Nano, mainly related to a wonky PyTorch installation.

As it stands, I'm getting on OOM memory error during inference on GPU. I haven't had time to devote to this project, but my next steps to make it work are:
- Re-train by fine tuning a smaller architecture, such as starting with a pretrained Resnet-50. This should be the path of least resistance because of the helpfully provided training code. To be completely honest, I'd be interested to investigate alternative architectures anyway in case a Resnet-101 is overfitting to this task.
- If I'm still getting an OOM error, try to compile to ONNYX. This looks like it could be made moderately more complicated because the existing model has a Python switch to decide between unrolled and sequential inference, and I don't believe ONNYX compilation in PyTorch supports this out of the box.

I'll let you know if/how I'm able to make it work.

Best,
Tim

**Comment by lojzezust** [comment]
Thanks, @playertr! I will update the code to include resizing by default, to avoid possible confusion with the input resolution.

Regarding the OOM and export. As you suggested, I encourage trying a more light-weight backbone such as ResNet-50. It should not be difficult to replace the current backbone. Let me know how it goes if you decide to do that. If you would like to contribute, I'll also be happy to include this version (and possibly its weights) in the repo.

Also make sure to run the network in **sequential** mode during inference, as this will increase the speed and reduce the memory consumption since only a single frame is processed at a time. 

If you decide to compile to ONNYX, it also makes sense to only compile the sequential version. I'm not sure if compiled models support internal buffers for storing features of the recent frames, but one workaround would be to implement this buffer outside the model and pass the recent features as an input to the model.

**Comment by playertr** [comment]
I was able to train up a MobileNetV3 really easily! It's a much smaller and faster network, and it seems to do pretty well on the easy examples. I have a link to the weights on my fork (https://github.com/playertr/WaSR-T) at the bottom of the README. It looks like the move to MobileNetV3 increases inferencespeed on my desktop to 42 FPS, halving the memory usage, while still providing reasonable results.

The code isn't ready to upstream because it makes sketchy assumptions about where to pull out the skip connections, and contains a number of hacks because I used different package versions.

I haven't had the time to put it on my Nano but I'm pretty sure it will not have the OOM errors anymore. I don't think I'll move to ONNYX if I can help it.

![tensorboard](https://user-images.githubusercontent.com/23692810/215638005-c33f9668-76cc-4c8f-8d82-3164fa216747.jpg)

Video results:
[wasrt_mobilenetv3.webm](https://user-images.githubusercontent.com/23692810/215637981-69ceb491-60b2-40e7-8316-4d86f5a96792.webm)

**Comment by irvinmarceloc** [comment]
hi everyone, are you using pytorch 2.0? I understand that jetson nano and compatible jetpack versions only support up to pytorch 1.8. I'm going to try this weekend to run the fork on the jetson nano, and report the results.

**Comment by playertr** [comment]
On the Nano, I was using the PyTorch 1.10 wheel from [here](https://qengineering.eu/install-pytorch-on-jetson-nano.html).

I'll try to test it on the Nano and make a legit mergeable PR before this weekend, but no guarantees.

**Comment by irvinmarceloc** [comment]
what happens when you run this command
```
python3
import torch
torch.cuda.is_available()
```
I was trying to install through pip but the result is False, it only works when I download one of the supported versions offered by nvidia

**Comment by playertr** [comment]
I get the following, now (I recently installed the NVIDIA-provided one, and the pytorch version shouldn't make a big difference as long as CUDA is available).

```
tim@sailboat:~/WaSR-T$ python3
Python 3.6.9 (default, Nov 25 2022, 14:10:45) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
True
>>> torch.__version__
'1.10.0'
```

**Comment by playertr** [comment]
I have an update on the training: I switched up my MobileNet implementation with a couple of changes:
1. I now take the skip connections from the first and second residual blocks of  MobileNet
2. I reduced the latent feature dimension of the feature fusion module from 1024 to 128
3. I upscaled the FFM output to 96x128 to increase the output resolution
4. I completed training for 500 epochs

These changes helped a lot with accuracy (changes 1, 3, and 4) and memory/latency (change 2). Now, the network does well enough on  the validation set that I'm concerned I may have inadvertently trained on validation examples. Alternatively, it could just be that MobileNetV3 happens to do really well at this task (supporting the ResNet101 overfitting hypothesis).

| Metric                    | Value  |
|---------------------------|--------|
| Val Accuracy              | 0.9996 |
| Val Obstacle IOU          | 0.9958 |
| Val Water IOU             | 0.9996 |
| Val Sky IOU               | 0.9994 |
| # Params                  | 18.0M  |
| RTX2060 Inf. Speed (FPS)  | 48     |
| Nano 4GB Inf. Speed (FPS) | 3.5    |

Unfortunately, inference on the Nano is much slower -- around 3.5 FPS. This is likely fine for many low-speed maritime applications (e.g., a sailboat avoiding a lighthouse). I was able to compile the network to ONNX and execute in a TensorRT environment, and saw no improvement in inference speed. As a side effect, because ONNX only readily supports directed acyclic graphs, I believe the compiled networks miss out on the temporal convolution benefits of WaSR-T because they always produce a deterministic output for a given frame regardless of temporal context.

I would not expect any higher Jetson Nano inference speed with this size of network because the FPS value is consistent with the advertised compute capability of each GPU: the RTX2060 has about 12 times as many GFLOPS as the Jetson Nano's chip, and executes about 12 times as fast. The only way to speed inference would be to reduce network size with pruning or quantization -- I believe these strategies were used to create the ~30 FPS nominal execution speed of other MobileNetV3 applications on the same hardware.

Anyway, I can package this up into a minimally invasive PR if you think it belongs despite the lackluster speed. I'm happy to let this languish in the issues and my fork if you prefer.

![wasrt_mobilenetv3](https://user-images.githubusercontent.com/23692810/216466600-6180120d-4518-46a4-a1f4-7d52e471ffc4.gif)
Fig. 1: Results on entire WasRT-1325 set (training + val partitions)

**Comment by lojzezust** [comment]
Great work, Tim! 

As you said, 3.5 FPS seems good enough for many applications, and at the very least it **does** run on the Jetson Nano. The speed could also be marginally further increased by reducing the number of context frames **T**.

I'd be happy to include your model in the repo if you prepare a PR. I believe we can add an `--architecture` or `--backbone` argument to the training and inference script to select which model to use (e.g. `wasrt_resnet101` or `wasrt_mobilenetv3`). We do something similar in the [WaSR repo](https://github.com/lojzezust/WaSR/blob/c9339632fa486ca83aad10719f9ff27a68c0ae36/train.py#L68).

I would also like to report the performance on the MODS benchmark, which might show if the method overfits to the training set. If you prepare a working PR and send me a link to the model weights I can run the inference and evaluation on my end to spare you the additional work, since I already have the pipeline set up.

**Comment by playertr** [comment]
Awesome, thank you for offering to check the MODS benchmark!

As a matter of fact, I just discovered that most Jetson Nano benchmarks perform inference at a quarter the resolution we had been using. By simply modifying the inference resolution using your handy `--resize` argument (and adding some questionably necessary `torch.jit.compile` invocations) the output FPS jumped to 15, which seems real-time enough for ICRA even though this is a just-for-fun project for me (by the way, see you in London this summer?).

I'm retraining on the 192x256 sized images right now, in case there is a difference in how the 3D convolution works with the downsampled images.

I'm also considering adding some additional code to support real-time inference from a gstreamer stream. Would you want that to be a part of the PR?

**Comment by lojzezust** [comment]
Great, I'm interested to see how well the model performs at such a low resolution.

Even though it's just a pet project, I'm sure that you could easily publish it at the top conferences/publications with some additional architectural changes to still allow quality predictions at a higher resolution and accuracy. By the way, we are also currently working on an embedded-compute-ready version of base WaSR. If you are interested I can let you know when it's released.

I won't be attending ICRA this year, but I hope I get to meet you at some future conferences.

Real-time inference support also sounds very useful. If it is possible to package your GStreamer code into a simple script alternative to `predict_sequential.py`, I would be happy to include it.

**Comment by playertr** [comment]
Thanks! I'll prepare the PR as soon as I can -- I hoped to get this done weeks ago, but life got in the way.

I look forward to crossing paths in marine robotics in the future! Please keep me posted on embedded WaSR.

**Comment by irvinmarceloc** [comment]
Hi everyone, I'm taking a long time to test the new architecture due to some hiccups. Try installing Pytorch versions 10.1.10 and 10.1.11 as indicated in this [link](https://qengineering.eu/install-pytorch-on-jetson-nano.html) . But I get this error

`AttributeError: Can't get attribute 'Conv2dNormActivation' on <module 'torchvision.ops.misc' from '/home/jettson/.local/lib/python3.6/site-packages/torchvision-0.9.0-py3.6-linux-aarch64.egg/torchvision/ops/misc.py'`

Then I was updated to pytorch 10.1.11 and its corresponding torchvision version but it doesn't work. How can I install version 2 of Pytorch on Jetson Nano?

What I can be doing wrong?

**Comment by playertr** [comment]
Hmm, it sounds like the version of pytorch you are using may incompatible with your torchvision installation. There should be a compatibility matrix somewhere in the pytorch docs. You might try `pip uninstall torchvision; pip install torchvision` to see if it fixes your installation by auto-installing a compatible version.

I don't think that the pytorch version ends up being important for this network. As long as you have a working, `torch.cuda.is_available()` version of pytorch compatible with CUDA, you should be ready to go. It's not worth upgrading to PyTorch 2.0. If it is possible to install the versions of pytorch-lightning, torchmetrics, and torchvision that are listed in `requirements.txt`, then you should do so because there are minor API changes in the latest versions of these packages. However, you might need to bump these packages' versions for compatibility with your main pytorch installation.

I'm working on my PR on a branch of my fork called "[mobile_inference](https://github.com/playertr/WaSR-T/tree/mobile_inference)". The bottom of the README has instructions to download the weights for the mobile-optimized model from Google Drive. The code seems to work on my x86 Linux desktop, but my next step is to install it on a totally fresh Jetson Nano and ensure the README and `requirements.txt` are set up for a seamless install before submitting the PR. Until I'm able to complete this there will likely be bugs; you're welcome to take this branch for a spin before that.

**Comment by lojzezust** [comment]
Closing. Addressed by #4. Feel free to open a new issue if you encounter problems with the new mobile version of the model.

**Comment by lojzezust** [comment]
Hi! We have recently released a new model eWaSR, specifically designed for running on devices with limited compute and memory. It achieves comparable performance to base WaSR at a fraction of its requirements. I think it would be a great fit for you guys.

If you are interested, feel free to check it out here: https://github.com/tersekmatija/eWaSR

**Comment by playertr** [comment]
That's awesome! What a great contribution, I'm excited to try it out.

Are you waiting on acceptance to put out an arxiv version? I'm keen to learn your reasoning for what you did differently.

**Comment by lojzezust** [comment]
Great, you can read the paper on arxiv: https://arxiv.org/abs/2304.11249

But I see that we didn't update the link in the repo yet though. Thanks for letting me know.
